tiledb/api/c_api/ndrectangle/ndrectangle_api_external_experimental.h

```diff
@@ -0,0 +1,200 @@
+ * copies of the Software, and to permit persons to whom the Software is
+ * furnished to do so, subject to the following conditions:
+ *
+ * The above copyright notice and this permission notice shall be included in
+ * all copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+ * THE SOFTWARE.
+ *
+ * @section DESCRIPTION
+ *
+ * This file declares the NDRectangle experimental C API for TileDB.
+ */
+
+#ifndef TILEDB_CAPI_NDRECTANGLE_API_EXTERNAL_EXPERIMENTAL_H
+#define TILEDB_CAPI_NDRECTANGLE_API_EXTERNAL_EXPERIMENTAL_H
+
+#include "tiledb/api/c_api/api_external_common.h"
+#include "tiledb/api/c_api/context/context_api_external.h"
+#include "tiledb/api/c_api/domain/domain_api_external.h"
+
+#ifdef __cplusplus
+extern "C" {
+#endif
+
+/**
+ * C API struct exposed to the user to help specify
+ * the limits of a dimension for a ND rectangle.
+ */
+typedef struct {
+  const void* min;
+  uint64_t min_size;
+  const void* max;
+  uint64_t max_size;
+} tiledb_range_t;
+
+typedef struct tiledb_ndrectangle_handle_t tiledb_ndrectangle_t;
+
+/**
+ * Allocate an N-dimensional rectangle given a TileDB array schema domain
+ * **Example:**
+ *
+ * @code{.c}
+ * tiledb_ndrectangle_t *ndr;
+ * tiledb_ndrectangle_alloc(ctx, domain, &ndr);
+ * tiledb_ndrectangle_free(&ndr);
+ * @endcode
+ *
+ * @param ctx The TileDB context
+ * @param domain The TileDB array schema domain
+ * @param ndr The n-dimensional rectangle to be allocated
+ * @return `TILEDB_OK` for success and `TILEDB_ERR` for error.
+ */
+TILEDB_EXPORT int32_t tiledb_ndrectangle_alloc(
+    tiledb_ctx_t* ctx,
+    tiledb_domain_t* domain,
+    tiledb_ndrectangle_t** ndr) TILEDB_NOEXCEPT;
+
+/**
+ * Free the resources associated with the N-dimensional rectangle arg
+ * **Example:**
+ *
+ * @code{.c}
+ * tiledb_ndrectangle_t *ndr;
+ * tiledb_ndrectangle_alloc(ctx, domain, &ndr);
+ * tiledb_ndrectangle_free(&ndr);
+ * @endcode
+ *
+ * @param ndr The n-dimensional rectangle to be freed
+ * @return `TILEDB_OK` for success and `TILEDB_ERR` for error.
+ */
+TILEDB_EXPORT int32_t tiledb_ndrectangle_free(tiledb_ndrectangle_t** ndr)
+    TILEDB_NOEXCEPT;
+
+/**
+ * Get the range set on an N-dimensional rectangle for a dimension name
+ *
+ * The pointers within the returned range struct point to resources managed
+ * by the `tiledb_ndrectangle_t` instance, it is not the responsibility of the
+ * caller to free those resources, attempting to do so results in undefined
+ * behavior.
```

Comment:
Consider adding that the resources are tied to the lifetime of the rectangle object.

This also applies to the next API.